presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotSegmentStreamingPageSource.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.pinot;
+
+import com.facebook.presto.common.Page;
+import com.facebook.presto.pinot.grpc.Constants;
+import com.facebook.presto.pinot.grpc.GrpcRequestBuilder;
+import com.facebook.presto.pinot.grpc.PinotStreamingQueryClient;
+import com.facebook.presto.pinot.grpc.ServerResponse;
+import com.facebook.presto.spi.ConnectorSession;
+import org.apache.pinot.common.utils.DataTable;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.presto.pinot.PinotErrorCode.PINOT_DATA_FETCH_EXCEPTION;
+import static com.facebook.presto.pinot.PinotErrorCode.PINOT_INVALID_SEGMENT_QUERY_GENERATED;
+import static com.facebook.presto.pinot.PinotErrorCode.PINOT_UNEXPECTED_RESPONSE;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This class retrieves Pinot data from a Pinot client, and re-constructs the data into Presto Pages.
+ */
+
+public class PinotSegmentStreamingPageSource
+        extends PinotSegmentPageSource
+{
+    private final PinotStreamingQueryClient pinotStreamingQueryClient;
+    private Iterator<ServerResponse> serverResponseIterator;
+    private long completedPositions;
+
+    public PinotSegmentStreamingPageSource(
+            ConnectorSession session,
+            PinotConfig pinotConfig,
+            PinotStreamingQueryClient pinotStreamingQueryClient,
+            PinotSplit split,
+            List<PinotColumnHandle> columnHandles)
+    {
+        super(session, pinotConfig, null, split, columnHandles);
+        this.pinotStreamingQueryClient = requireNonNull(pinotStreamingQueryClient, "pinotStreamingQueryClient is null");
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return closed;
+    }
+
+    /**
+     * @return constructed page for pinot data.
+     */
+    @Override
+    public Page getNextPage()
+    {
+        if (closed) {
+            return null;
+        }
+        if (serverResponseIterator == null) {
+            serverResponseIterator = queryPinot(split);
+        }
+        ByteBuffer byteBuffer = null;
+        if (serverResponseIterator.hasNext()) {
```

Comment:
Can you add a small comment here that we expect to see a sequence of Data followed by Metadata ?

Reply:
will do